lib/http_aws_sigv4.c

```diff
@@ -66,6 +70,188 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
+                             struct dynbuf *signed_headers)
+{
+  char *date_hdr_key =  curl_maprintf("X-%s-Date", provider1_mid);
+  char *date_full_hdr = curl_maprintf("x-%s-date:%s",
+                                      provider1_low, timestamp);
+  struct curl_slist *head = NULL;
+  struct curl_slist *tmp_head = NULL;
+  int ret = CURLE_OUT_OF_MEMORY;
+  struct curl_slist *l;
+  int again = 1;
+  char *full_host = NULL;
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  if(Curl_checkheaders(data, STRCONST("Host"))) {
+    head = NULL;
+  }
+  else {
+    if(data->state.aptr.host) {
+      size_t pos;
+
+      full_host = strdup(data->state.aptr.host);
+      if(!full_host)
+        goto fail;
+      /* remove /r/n as the separator for canonical request must be '\n'
+       * and are readd after */
+      pos = strlen(full_host);
+      if(pos > 1 && full_host[pos - 1] == '\n') {
+        full_host[pos - 1] = 0;
+        if(pos > 2 && full_host[pos - 2] == '\r')
+          full_host[pos - 2] = 0;
+      }
+    }
+    else {
+      full_host = curl_maprintf("host:%s", hostname);
+    }
+
+    if(!full_host)
+      goto fail;
+    head = curl_slist_append(NULL, full_host);
+    if(!head)
+      goto fail;
+  }
+
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  for(l = data->set.headers; l; l = l->next) {
+    tmp_head = curl_slist_append(head, l->data);
+    if(!tmp_head)
+      goto fail;
+    head = tmp_head;
+  }
+
+  /* remove whitespace, and lowercase each headers */
+  for(l = head; l; l = l->next) {
+    char *tmp = l->data, *tmp2;
+    size_t pos;
+
+    for(pos = 0; tmp[pos] && tmp[pos] != ':'; ++pos);
+    Curl_strntolower(l->data, l->data, pos);
+
+    tmp += pos;
+    if(!*tmp)
+      continue;
+    ++tmp;
+    tmp2 = tmp;
+    while(*tmp2 == ' ' || *tmp2 == '\t') {
+      ++tmp2;
+    }
+    if(tmp != tmp2)
+      memmove(tmp, tmp2, strlen(tmp2) + 1);
+
+    while(*tmp) {
+      tmp2 = tmp;
+      while(*tmp2 == ' ' || *tmp2 == '\t') {
+        ++tmp2;
+      }
+
+      if(tmp != tmp2) {
+        if(*tmp2 == 0)
+          /* remove spaces at the end */
+          *tmp = 0;
+        else {
+          *tmp++ = ' '; /* if tab convert to space */
+          if(*tmp == ' ' || *tmp == '\t')
+            memmove(tmp, tmp2, strlen(tmp2));
+        }
+        tmp = tmp2;
+      }
+      else {
+        ++tmp;
+      }
+    }
+  }
+
+  *date_header = find_date_hdr(data, date_hdr_key);
+  if(!*date_header) {
+    tmp_head = curl_slist_append(head, date_full_hdr);
+    if(!tmp_head)
+      goto fail;
+    head = tmp_head;
+    *date_header = curl_maprintf("%s: %s", date_hdr_key, timestamp);
+  }
+  else {
+    char *tmp;
+
+    *date_header = strdup(*date_header);
+    tmp = strchr(*date_header, ':');
+    if(!tmp)
+      goto fail;
+    ++tmp;
+    while(*tmp == ' ' || *tmp == '\t')
+      ++tmp;
+    strncpy(timestamp, tmp, TIMESTAMP_SIZE - 1);
+    timestamp[TIMESTAMP_SIZE - 1] = 0;
+  }
+
+  /* reorder */
```

Comment:
And by this you mean you alpha-sort them in a case sensitive manner, right?

Not *terribly* important but the reason libcurl sends Host: as the first header is this paragraph from [RFC 7230 section 3.2.2](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.2):

>   it is good practice to send  header fields that contain control data first, such as Host on  requests and Date on responses, so that implementations can decide  when not to handle a message as early as possible.

Reply:
Yes,
Note that I don't reorder the header send by curl (if I do, then my code is buggy)
I create a temporary string that contain the apha sorted headers, so I can create an hash out of the string.